src/GitHub.App/Services/GistCreator.cs

```diff
@@ -0,0 +1,76 @@
+﻿using System;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using System.Reactive.Linq;
+using System.Reactive.Threading.Tasks;
+using System.Threading.Tasks;
+using GitHub.Api;
+using GitHub.Extensions;
+using GitHub.Models;
+using GitHub.UI;
+using Octokit;
+using GitHub.Factories;
+using GitHub.Primitives;
+
+namespace GitHub.Services
+{
+    [Export(typeof(IGistCreator))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    public class GistCreator : IGistCreator
+    {
+        readonly ISelectedTextProvider selectedTextProvider;
+        readonly IConnectionManager connectionManager;
+        readonly IRepositoryHosts repositoryHosts;
+        readonly IUIProvider uiProvider;
+        readonly IApiClient apiClient;
+
+        [ImportingConstructor]
+        public GistCreator(ISelectedTextProvider selectedTextProvider, IConnectionManager connectionManager, 
+            IRepositoryHosts repositoryHosts, IUIProvider uiProvider, IApiClientFactory apiClientFactory)
+        {
+            this.selectedTextProvider = selectedTextProvider;
+            this.connectionManager = connectionManager;
+            this.repositoryHosts = repositoryHosts;
+            this.uiProvider = uiProvider;
+            this.apiClient = apiClientFactory.Create(HostAddress.GitHubDotComHostAddress);
+        }
+
+        public async Task<Gist> CreateFromSelectedText()
+        {
+            await EnsureLoggedIn();
+
+            var selectedText = selectedTextProvider.GetSelectedText().ToTask().Result;
+
+            // Show the Create Gist Window, and on Create button clicked execute the following
+            return await apiClient.CreateGist("NameWillBeEnteredInThePopup", true, selectedText)
```

Comment:
Yeah, we're a bit limited in some areas... This is why there's a big split in the Exports and UI libraries, and also why there's an ApiClient implementation in `GitHub.App` and a separate api library in `GitHub.Api`... `GitHub.Api`, `GitHub.UI`, `GitHub.Exports` and `GitHub.Extensions` are loaded up front as soon as VS and Team Explorer load, and they have the minimum dependencies required to do things. `GitHub.App`, `GitHub.Exports.Reactive`, and `GitHub.UI.Reactive` load when other things get called - namely the popup dialogs, which are "on demand" when the user does some specific action/request.

Unless something is needed in the home page, it's best to avoid adding things to the constrained libraries, and just add them to the non-constrained ones instead.